### PR TITLE
RandomRotation and fill

### PR DESCRIPTION
**Pull request by tonystratum**
Originally (vision/nightly), when `torchvision.transforms.RandomRotation` is used with a Tensor of shape `(C, H, W)`, without a `fill` argument, both in a `torch.nn.Sequential()`:
```python
tensor = torch.rand(3, 224, 224)

transforms = torch.nn.Sequential(
    torchvision.transforms.Resize((224,224)),
    torchvision.transforms.Normalize(mean=[0.485, 0.456, 0.406],
                                   std=[0.229, 0.224, 0.225]),
    torchvision.transforms.RandomRotation(45)
)

transformed = transforms(tensor)
```

we get a following error:
<details>
  <summary>Error</summary>
 
  
  ```python
Traceback (most recent call last):
  File "repr.py", line 13, in <module>
    transformed = transforms(tensor)
  File "/home/stratum/miniconda3/envs/pytorch-cpu-nightly/lib/python3.6/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/stratum/miniconda3/envs/pytorch-cpu-nightly/lib/python3.6/site-packages/torch/nn/modules/container.py", line 119, in forward
    input = module(input)
  File "/home/stratum/miniconda3/envs/pytorch-cpu-nightly/lib/python3.6/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/stratum/miniconda3/envs/pytorch-cpu-nightly/lib/python3.6/site-packages/torchvision/transforms/transforms.py", line 1236, in forward
    fill = [float(f) for f in fill]
TypeError: 'NoneType' object is not iterable
  ```
  
</details>

and by itself:
```python
tensor = torch.rand(3, 224, 224)

tensor = torchvision.transforms.Resize((224,224))(tensor)
tensor = torchvision.transforms.Normalize(mean=[0.485, 0.456, 0.406],
        std=[0.229, 0.224, 0.225])(tensor)
tensor = torchvision.transforms.RandomRotation(45)(tensor)
```

<details>
  <summary>Error</summary>
 
  
  ```python
Traceback (most recent call last):
  File "repr1.py", line 9, in <module>
    tensor = torchvision.transforms.RandomRotation(45)(tensor)
  File "/home/stratum/miniconda3/envs/pytorch-cpu-nightly/lib/python3.6/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/stratum/miniconda3/envs/pytorch-cpu-nightly/lib/python3.6/site-packages/torchvision/transforms/transforms.py", line 1236, in forward
    fill = [float(f) for f in fill]
TypeError: 'NoneType' object is not iterable
  ```
  
</details>

The fix is to check, if `img` is a `Tensor`, and if `fill` is `None`, set it to zero.

Unit tests:

<details>
  <summary>pytest test/test_transforms.py -vvv -k test_rotate</summary>
 
  
  ```python
=================================== test session starts ====================================
platform linux -- Python 3.8.5, pytest-6.2.2, py-1.10.0, pluggy-0.13.1 -- /home/stratum/PycharmProjects/venvs/tv-test/bin/python
cachedir: .pytest_cache
rootdir: /home/stratum/build/myvision/vision
collected 70 items / 68 deselected / 2 selected

test/test_transforms.py::Tester::test_rotate PASSED                                  [ 50%]
test/test_transforms.py::Tester::test_rotate_fill PASSED                             [100%]

===================================== warnings summary =====================================
torchvision/__init__.py:26
  /home/stratum/build/myvision/vision/torchvision/__init__.py:26: UserWarning: You are importing torchvision within its own root folder (/home/stratum/build/myvision/vision). This is not expected to work and may give errors. Please exit the torchvision project source and relaunch your python interpreter.
    warnings.warn(message.format(os.getcwd()))

-- Docs: https://docs.pytest.org/en/stable/warnings.html
======================= 2 passed, 68 deselected, 1 warning in 0.41s ========================
  ```
  
</details>

<details>
  <summary>pytest test/test_transforms_tensor.py -vvv -k test_random_rotate</summary>
 
  
  ```python
=================================== test session starts ====================================
platform linux -- Python 3.8.5, pytest-6.2.2, py-1.10.0, pluggy-0.13.1 -- /home/stratum/PycharmProjects/venvs/tv-test/bin/python
cachedir: .pytest_cache
rootdir: /home/stratum/build/myvision/vision
collected 56 items / 54 deselected / 2 selected

test/test_transforms_tensor.py::Tester::test_random_rotate PASSED                    [ 50%]
test/test_transforms_tensor.py::CUDATester::test_random_rotate SKIPPED (Skip if ...) [100%]

===================================== warnings summary =====================================
torchvision/__init__.py:26
  /home/stratum/build/myvision/vision/torchvision/__init__.py:26: UserWarning: You are importing torchvision within its own root folder (/home/stratum/build/myvision/vision). This is not expected to work and may give errors. Please exit the torchvision project source and relaunch your python interpreter.
    warnings.warn(message.format(os.getcwd()))

-- Docs: https://docs.pytest.org/en/stable/warnings.html
================== 1 passed, 1 skipped, 54 deselected, 1 warning in 2.92s ==================
  ```
  
</details>

flake8 and mypy passed.

Hope this helps!

### CODE REVIEW

**Comment by voldemortX** [comment]
In summary of #3309, please not convert fill, but:
1. set default as fill=0 for all places. 
2. clearly state in the docstrings that fill is no longer optional, only defaulted as 0, and check it to be a Sequence or Number.
3. probably no need adding tests.

File: torchvision/transforms/transforms.py [hunk 1]
```diff
@@ -692,6 +692,10 @@ def __init__(self, distortion_scale=0.5, p=0.5, interpolation=InterpolationMode.
 
         self.interpolation = interpolation
         self.distortion_scale = distortion_scale
+
+        if not isinstance(fill, Sequence) and not isinstance(fill, numbers.Number):
```
Comment (voldemortX) [inline torchvision/transforms/transforms.py:696]: maybe `isinstance(fill, (Sequence, numbers.Number))`?

Comment (tonystratum) [reply torchvision/transforms/transforms.py]: makes sense, pushing a fix

Comment (tonystratum) [reply torchvision/transforms/transforms.py]: @voldemortX 0fcb13975bb3e8ff199ba65060051ae7000df44a

**Comment by datumbox** [comment]
@tonystratum Thanks a lot for the PR! The changes seem reasonable to me but I'll let @vfdev-5 to approve because of your discussion at #3309.

@voldemortX Thanks for all your help on reviews and for providing good input.

File: torchvision/transforms/transforms.py [hunk 2]
```diff
@@ -692,6 +692,10 @@ def __init__(self, distortion_scale=0.5, p=0.5, interpolation=InterpolationMode.
 
         self.interpolation = interpolation
         self.distortion_scale = distortion_scale
+
+        if not isinstance(fill, (Sequence, numbers.Number)):
```
Comment (vfdev-5) [inline torchvision/transforms/transforms.py:696]: Maybe, we can keep BC by doing:
```python
if fill is None:
    fill = 0
```
Sorry, I'm a bit hesitating about introducing new behaviour.

File: torchvision/transforms/transforms.py [hunk 3]
```diff
@@ -1212,6 +1216,10 @@ def __init__(
 
         self.resample = self.interpolation = interpolation
         self.expand = expand
+
+        if not isinstance(fill, (Sequence, numbers.Number)):
```
Comment (vfdev-5) [inline torchvision/transforms/transforms.py:1220]: Same here

File: torchvision/transforms/transforms.py [hunk 4]
```diff
@@ -1234,7 +1242,9 @@ def forward(self, img):
         """
         fill = self.fill
         if isinstance(img, Tensor):
-            if isinstance(fill, (int, float)):
+            if fill is None:
```
Comment (vfdev-5) [inline torchvision/transforms/transforms.py:1245]: If we set fill as 0 in all the cases previously, we can skip this

**Review by vfdev-5 — COMMENTED** [review]
Thanks a lot @tonystratum ! 
I left few comments. Sorry, I think, it would be better to set fill=0 if it was passed as None...

File: torchvision/transforms/transforms.py (hunk 2 above)
Comment (tonystratum) [reply torchvision/transforms/transforms.py:696]: Agreed. Maybe throw a warning with it or something?

Comment (tonystratum) [reply torchvision/transforms/transforms.py:696]: @vfdev-5

Comment (vfdev-5) [reply torchvision/transforms/transforms.py:696]: In some sense it was already done by PIL in `_parse_fill*` method, so let's just reassign it without a warning.

**Review by vfdev-5 — COMMENTED** [review]
@tonystratum thanks for the update ! Could you please also include a test here: https://github.com/pytorch/vision/blob/master/test/test_transforms.py#L1527

Something like initialize a transform and verfiy then that its attribute is correctly set.
Also, for affine and perspective ops. Thanks a lot!

**Comment by tonystratum** [comment]
<details>
  <summary>unit test reports</summary>
 
  
  ```python
============================================== test session starts ===============================================
platform linux -- Python 3.8.5, pytest-6.2.2, py-1.10.0, pluggy-0.13.1 -- /home/stratum/PycharmProjects/venvs/tv-test2/bin/python
cachedir: .pytest_cache
rootdir: /home/stratum/build/myvision/vision
collected 70 items / 69 deselected / 1 selected

test/test_transforms.py::Tester::test_random_rotation PASSED                                               [100%]

================================================ warnings summary ================================================
torchvision/__init__.py:26
  /home/stratum/build/myvision/vision/torchvision/__init__.py:26: UserWarning: You are importing torchvision within its own root folder (/home/stratum/build/myvision/vision). This is not expected to work and may give errors. Please exit the torchvision project source and relaunch your python interpreter.
    warnings.warn(message.format(os.getcwd()))

-- Docs: https://docs.pytest.org/en/stable/warnings.html
================================== 1 passed, 69 deselected, 1 warning in 0.30s ===================================
  ```

```python
============================================== test session starts ===============================================
platform linux -- Python 3.8.5, pytest-6.2.2, py-1.10.0, pluggy-0.13.1 -- /home/stratum/PycharmProjects/venvs/tv-test2/bin/python
cachedir: .pytest_cache
rootdir: /home/stratum/build/myvision/vision
collected 70 items / 69 deselected / 1 selected

test/test_transforms.py::Tester::test_random_affine PASSED                                                 [100%]

================================================ warnings summary ================================================
torchvision/__init__.py:26
  /home/stratum/build/myvision/vision/torchvision/__init__.py:26: UserWarning: You are importing torchvision within its own root folder (/home/stratum/build/myvision/vision). This is not expected to work and may give errors. Please exit the torchvision project source and relaunch your python interpreter.
    warnings.warn(message.format(os.getcwd()))

-- Docs: https://docs.pytest.org/en/stable/warnings.html
================================== 1 passed, 69 deselected, 1 warning in 0.31s ===================================
```

```python
============================================== test session starts ===============================================
platform linux -- Python 3.8.5, pytest-6.2.2, py-1.10.0, pluggy-0.13.1 -- /home/stratum/PycharmProjects/venvs/tv-test2/bin/python
cachedir: .pytest_cache
rootdir: /home/stratum/build/myvision/vision
collected 70 items / 68 deselected / 2 selected

test/test_transforms.py::Tester::test_randomperspective PASSED                                             [ 50%]
test/test_transforms.py::Tester::test_randomperspective_fill PASSED                                        [100%]

================================================ warnings summary ================================================
torchvision/__init__.py:26
  /home/stratum/build/myvision/vision/torchvision/__init__.py:26: UserWarning: You are importing torchvision within its own root folder (/home/stratum/build/myvision/vision). This is not expected to work and may give errors. Please exit the torchvision project source and relaunch your python interpreter.
    warnings.warn(message.format(os.getcwd()))

-- Docs: https://docs.pytest.org/en/stable/warnings.html
================================== 2 passed, 68 deselected, 1 warning in 0.32s ===================================
```
  
</details>

There was a dedicated `test_randomperspective_fill` for some reason, so I decided to put our tests there.

cc @vfdev-5

File: test/test_transforms.py [hunk 5]
```diff
@@ -180,6 +180,14 @@ def test_randomperspective(self):
                                torch.nn.functional.mse_loss(tr_img2, F.to_tensor(img)))
 
     def test_randomperspective_fill(self):
+
+        # assert fill being either a Sequence or a Number
+        with self.assertRaises(TypeError):
+            transforms.RandomPerspective(fill=NotImplemented)
```
Comment (vfdev-5) [inline test/test_transforms.py:186]: nit: let's replace `NotImplemented` by `{}` or something else. Seems like NotImplemented has a special meaning for python: https://docs.python.org/3/library/constants.html#NotImplemented

**Review by vfdev-5 — COMMENTED** [review]
Thanks @tonystratum, looks good ! I left a comment about NotImplemented.

File: test/test_transforms.py [hunk 6]
```diff
@@ -183,7 +183,7 @@ def test_randomperspective_fill(self):
 
         # assert fill being either a Sequence or a Number
         with self.assertRaises(TypeError):
-            transforms.RandomPerspective(fill=NotImplemented)
+            transforms.RandomPerspective(fill={})
```
Comment (tonystratum) [inline test/test_transforms.py:186]: @vfdev-5

**Comment by tonystratum** [comment]
> formatting problem

Yeah, sorry for that

**Review by vfdev-5 — APPROVED** [review]
Looks good so far, thanks @tonystratum !

**Comment by tonystratum** [comment]
@vfdev-5 Thanks for your time!)

**Pull request merged**